### PR TITLE
Antispam Fix

### PR DESCRIPTION
**Pull request by Kingwolf123**
Updated Antispam Plugin to work with newer API.
Added Admin Authentication to exempt admins from being monitored (optional)
Added ReadMe for Cvars Table.

### CODE REVIEW

File: plugins/antispam/ReadMe.md [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+#
+#
+
+CVAR | DESCRIPTION
+------------- | --------------------------
+__antispam_maxMessagesPerMinute__ | *Count of maximum messages a player can send in a minute. 0 disables the chat completely. Value Range : (0 - 30)*
+__antispam_minAdminPower__ | *Minimum power points which disable the spam-control. 0 means enabled for everyone. Value Range : (0 - 100)*
+__antispam_minMessageDelay__ | *Amount of time after sending a message after which the player can chat again. 0 disables the limit. Value Range : (0 - 60)*
+__antispam_renewedMessageDelay__ | *Penalize players and add penalty time if player chats before minMessageDelay. Value Range : (0 - 1)*
+
```
Comment (T-Maxxx) [inline plugins/antispam/ReadMe.md]: @Kingwolf123 If there was supposed to be indentation... Well... you tried ;D

Comment (T-Maxxx) [reply plugins/antispam/ReadMe.md]: I think this must be merged and whole plugin code should be moved to a seperate repository.

Comment (Kingwolf123) [reply plugins/antispam/ReadMe.md]: Haha. I added extra lines in between the header of MD and the table by mistake. Not big on markdown.

File: plugins/antispam/antispam_plugin.cpp [hunk 2]
```diff
@@ -1,14 +1,33 @@
...
+ 
 #include "../pinc.h"
+#include <chrono> // for time_t typdef
 
-#define ANTISPAM_MAXMESSAGES 30
+using namespace std;
 
-typedef struct{
+PCL int OnInit();
+#define max 8
```
Comment (D4edalus) [inline plugins/antispam/antispam_plugin.cpp]: this is a horrible name for a preprocessor definition, please change back to ANTISPAM_MAXMESSAGES

File: plugins/antispam/antispam_plugin.cpp [hunk 3]
```diff
@@ -18,37 +37,50 @@ typedef struct{
...
 
-PCL int OnInit(){	// Funciton called on server initiation
+PCL int OnInit()   // Function called on server initiation
+{	
 
-	//G_SayCensor_Init();
 	data.maxPlayers = Plugin_GetSlotCount();
 	data.maxMPM = Plugin_Cvar_RegisterFloat("antispam_maxMessagesPerMinute",8,0,30,0,"Count of maximum messages a player can send in a minute. 0 disables the chat completely.");
-	data.minAP = Plugin_Cvar_RegisterInt("antispam_minAdminPower",50,0,100,0,"Minimum power points which disable the player. 0 means enabled for everyone.");
+	data.minAP = Plugin_Cvar_RegisterInt("antispam_minAdminPower",50,0,100,0,"Minimum power points which disable the spam-control. 0 means enabled for everyone.");
```
Comment (D4edalus) [inline plugins/antispam/antispam_plugin.cpp]: should be rather: Minimum power points to disable spam-control.

File: plugins/antispam/antispam_plugin.cpp [hunk 4]
```diff
@@ -95,10 +134,10 @@ PCL void OnInfoRequest(pluginInfo_t *info){	// Function used to obtain informati
...
 
     // =====  OPTIONAL  FIELDS  =====
-    info->pluginVersion.major = 1;
-    info->pluginVersion.minor = 0;	// Plugin version
-    strncpy(info->fullName,"IceOps antispam plugin by TheKelm",sizeof(info->fullName)); //Full plugin name
-    strncpy(info->shortDescription,"This plugin is used to prevent spam in the ingame chat.",sizeof(info->shortDescription)); // Short plugin description
-    strncpy(info->longDescription,"This plugin is used to prevent spam in the ingame chat. To personalize the settings, set corresponding cvars. Copyright (c) 2013 IceOps",sizeof(info->longDescription));
+    info->pluginVersion.major = 2;
+    info->pluginVersion.minor = 1;	// Plugin version
+    strncpy(info->fullName,"Cod4X Antispam plugin by Usmania",sizeof(info->fullName)); //Full plugin name
```
Comment (D4edalus) [inline plugins/antispam/antispam_plugin.cpp]: the plugin was not created by you. editing a handful of lines doesnt make you the creator.

Comment (Kingwolf123) [reply plugins/antispam/antispam_plugin.cpp]: Oh yeah my bad. I rewrote the whole thing just to get the hang of the logic and different Functions.
I will change it back.

**Comment by Kingwolf123** [comment]
@D4edalus Check

File: plugins/antispam/antispam_plugin.cpp [hunk 5]
```diff
@@ -58,35 +90,42 @@ PCL void OnMessageSent(char *message,int slot, qboolean *show, int type){
...
 	}
-	Plugin_ChatPrintf(slot,"Debug: %d, %d, %d, %d.\n",j,i,data.maxMPM->integer,data.maxMPM->latchedinteger);
 	for(i=j;i<ANTISPAM_MAXMESSAGES;++i){
 	    data.players[slot].messages[i] = 0;
 	}
 
 	// j now holds the value of how many messages this player has sent in the last 60 seconds. If it is too much - dont show the msg
 	// if it is ok - show the message and add time to messages[j]
-	if(j<data.maxMPM->integer){
+	if(j<ANTISPAM_MAXMESSAGES)
```
Comment (D4edalus) [inline plugins/antispam/antispam_plugin.cpp]: what's the point of replacing maxMPM with ANTISPAM_MAXMESSAGES? you just make the number of message unconfigurable that way. you even documented maxMPM in your readme, but now it is useless.

**Comment by Kingwolf123** [comment]
@D4edalus Sorry didn't have time earlier. The entire plugin has almost been revamped with a new algorithm.
Check now things should work smoothly now.

**Pull request merged**